File: src/FixWPBouncing.cs
/*
    Cordova Fix-WP-Bouncing Plugin
    https://github.com/vilic/cordova-plugin-fix-wp-bouncing
    
    by VILIC VANE
    https://github.com/vilic

    MIT License
*/

using LinqToVisualTree;
using Microsoft.Phone.Controls;
using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using WPCordovaClassLib;
using WPCordovaClassLib.Cordova.Commands;

namespace Cordova.Extension.Commands {
    class FixWPBouncing : BaseCommand {
        WebBrowser browser;

        public FixWPBouncing() {
            var frame = Application.Current.RootVisual as PhoneApplicationFrame;
            var page = frame.Content as PhoneApplicationPage;
            var grid = page.FindName("LayoutRoot") as Grid;
            var cordovaView = grid.FindName("CordovaView") as CordovaView;
            var cordovaViewGrid = cordovaView.FindName("LayoutRoot") as Grid;

            browser = cordovaViewGrid.FindName("CordovaBrowser") as WebBrowser;

            var border = browser.Descendants<Border>().Last() as Border;

            border.ManipulationDelta += border_ManipulationDelta;
            border.ManipulationCompleted += border_ManipulationCompleted;
        }

        void border_ManipulationCompleted(object sender, ManipulationCompletedEventArgs e) {
            browser.InvokeScript("eval", "FixWPBouncing.onmanipulationcompleted()");
        }

        void border_ManipulationDelta(object sender, ManipulationDeltaEventArgs e) {
            if (e.DeltaManipulation.Translation.Y != 0) {
                var status = browser.InvokeScript("eval", "FixWPBouncing.onmanipulationdelta()") as string;
                
                if (e.DeltaManipulation.Translation.Y > 0) {
                    if (status == "top" || status == "both") {
                        e.Handled = true;
                    }
                } else {
                    if (status == "bottom" || status == "both") {
                        e.Handled = true;
                    }
                }
            }
        }
    }
}

File: www/fix-wp-bouncing.js
/*

    Cordova Fix-WP-Bouncing Plugin
    https://github.com/vilic/cordova-plugin-fix-wp-bouncing

    by VILIC VANE
    https://github.com/vilic

    MIT License

*/

var POINTER_DOWN;

if (navigator.pointerEnabled) {
    POINTER_DOWN = 'pointerdown';
} else if (navigator.msPointerEnabled) {
    POINTER_DOWN = 'MSPointerDown';
}

exports.target = null;

exports.onmanipulationdelta = function () {
    if (!exports.target) {
        return '';
    }

    var target = exports.target;

    var top = target.scrollTop == 0;
    var bottom = target.scrollTop + target.clientHeight == target.scrollHeight;

    return top ? bottom ? 'both' : 'top': bottom ? 'bottom' : '';
};

exports.onmanipulationcompleted = function () {
    exports.target = null;
};

exports.fix = function (target) {
    if (!POINTER_DOWN) {
        return;
    }

    if (!(target instanceof HTMLElement) && target.length) {
        target = target[0];
    }

    target.addEventListener(POINTER_DOWN, function () {
        exports.target = target;
    }, false);
};